File: 3. Behavioral Design Patterns/Memento/C#/MementoExample.cs
// Memento Pattern Example in C#
// Text editor that can save and restore document states

using System;
using System.Collections.Generic;

namespace MementoExample
{
    class TextMemento
    {
        public string Content { get; }
        public TextMemento(string content) { Content = content; }
    }

    class TextEditor
    {
        private string content = "";
        public void Write(string text) { content += text; }
        public void Clear() { content = ""; }
        public string GetContent() => content;
        public TextMemento Save() => new TextMemento(content);
        public void Restore(TextMemento memento) { content = memento.Content; }
    }

    class EditorHistory
    {
        private readonly Stack<TextMemento> history = new();
        public void Save(TextEditor editor) { history.Push(editor.Save()); }
        public void Undo(TextEditor editor) { if (history.Count > 0) editor.Restore(history.Pop()); }
    }

    class Program
    {
        static void Main()
        {
            var editor = new TextEditor();
            var history = new EditorHistory();

            editor.Write("Hello ");
            history.Save(editor);
            Console.WriteLine("Content: " + editor.GetContent());

            editor.Write("World!");
            history.Save(editor);
            Console.WriteLine("Content: " + editor.GetContent());

            history.Undo(editor);
            Console.WriteLine("After undo: " + editor.GetContent());

            history.Undo(editor);
            Console.WriteLine("After undo: " + editor.GetContent());
        }
    }
}

File: 3. Behavioral Design Patterns/Memento/Javascript/MementoExample.js
// Memento Pattern Example in JavaScript
// Text editor that can save and restore document states

class TextMemento {
  constructor(content) { this.content = content }
  getContent() { return this.content }
}

class TextEditor {
  constructor() { this.content = '' }
  write(text) { this.content += text }
  clear() { this.content = '' }
  getContent() { return this.content }
  save() { return new TextMemento(this.content) }
  restore(memento) { this.content = memento.getContent() }
}

class EditorHistory {
  constructor() { this.history = [] }
  save(editor) { this.history.push(editor.save()) }
  undo(editor) { if (this.history.length > 0) editor.restore(this.history.pop()) }
}

;(function main() {
  const editor = new TextEditor()
  const history = new EditorHistory()

  editor.write('Hello ')
  history.save(editor)
  console.log('Content:', editor.getContent())

  editor.write('World!')
  history.save(editor)
  console.log('Content:', editor.getContent())

  history.undo(editor)
  console.log('After undo:', editor.getContent())

  history.undo(editor)
  console.log('After undo:', editor.getContent())
})()
